coresimd/acle/hints.rs

```diff
@@ -0,0 +1,115 @@
+// # References
+//
+// - Section 7.4 "Hints" of ACLE
+// - Section 7.7 "NOP" of ACLE
+
+/// Generates a WFI (wait for interrupt) hint instruction, or nothing.
+///
+/// The WFI instruction allows (but does not require) the processor to enter a
+/// low-power state until one of a number of asynchronous events occurs.
+// Section 10.1 of ACLE says that the supported arches are: 8, 6K, 6-M
+// LLVM says "instruction requires: armv6k"
+#[cfg(any(target_feature = "v6k", target_arch = "aarch64"))]
+#[inline(always)]
+pub unsafe fn __wfi() {
+    asm!("WFI" : : : : "volatile")
+}
+
+/// Generates a WFE (wait for event) hint instruction, or nothing.
+///
+/// The WFE instruction allows (but does not require) the processor to enter a
+/// low-power state until some event occurs such as a SEV being issued by
+/// another processor.
+// Section 10.1 of ACLE says that the supported arches are: 8, 6K, 6-M
+// LLVM says "instruction requires: armv6k"
+#[cfg(any(target_feature = "v6k", target_arch = "aarch64"))]
+#[inline(always)]
+pub unsafe fn __wfe() {
+    asm!("WFE" : : : : "volatile")
+}
+
+/// Generates a SEV (send a global event) hint instruction.
+///
+/// This causes an event to be signaled to all processors in a multiprocessor
+/// system. It is a NOP on a uniprocessor system.
+// Section 10.1 of ACLE says that the supported arches are: 8, 6K, 6-M, 7-M
+// LLVM says "instruction requires: armv6k"
+#[cfg(any(target_feature = "v6k", target_arch = "aarch64"))]
+#[inline(always)]
+pub unsafe fn __sev() {
+    asm!("SEV" : : : : "volatile")
+}
+
+/// Generates a send a local event hint instruction.
+///
+/// This causes an event to be signaled to only the processor executing this
+/// instruction. In a multiprocessor system, it is not required to affect the
+/// other processors.
+// LLVM says "instruction requires: armv8"
+#[cfg(target_arch = "aarch64")]
```

Comment:
What's the cfg for 32-bit ARMv8? There's no target_arch = "aarch32" and AFAICT  we don't have any 32-bit v8-a target listed in `rustc --print target-list`

llvm / rustc seem to be OK with emitting a SEVL instruction for thumbv7m-none-eabi and thumbv8m.main-none-eabi though

Reply:
You can use the "v8" target feature, it works the same way as the "v7" feature.

LLVM treats SEVL as an alias for the HINT instruction (`HINT #5`) and doesn't actually check if the given hint code is supported by the current target. However this is mostly harmless since unrecognized hints are treated as NOPs by the architecture.

Reply:
Done. We'll need to whitelist the v8 on arch=aarch64. I don't know if that feature is ever enabled on arch=arm -- it's not set for `thumbv8m.*-none-eabi*`; I would have to check if it's set for ARMv8-R.

Reply:
> it (v8) is not set for `thumbv8m.*-none-eabi*`

I should add that *is* correct; there's no SEVL on ARMv8-M.

> I would have to check if it's set for ARMv8-R

I checked with an hypothetical armv8r-none-eabi target and yes v8 is set. SEVL is available on ARMv8-R so this makes sense.

``` console
$ rustc +master --print cfg --target armv8r-none-eabi.json | grep feature
target_feature="dsp"
target_feature="neon"
target_feature="rclass"
target_feature="v5te"
target_feature="v6"
target_feature="v6k"
target_feature="v6t2"
target_feature="v7"
target_feature="v8"
target_feature="vfp2"
target_feature="vfp3"
target_feature="vfp4"
```

Reply:
> We'll need to whitelist the v8 on arch=aarch64

Err, actually `+v8` is not a valid aarch64 feature. It doesn't show up in `llc -march=aarch64 -mattr=help`. There's `+v8.1a` and others but they are not set by default for the aarch64-unknown-linux-gnu target.

Reply:
I have updated the cfg to include arch=aarch64.